File: app/src/main/java/app/trigger/ImageActivity.java
package app.trigger;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.util.Base64;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import java.io.ByteArrayOutputStream;


public class ImageActivity extends AppCompatActivity {
    private static final int READ_IMAGE_REQUEST = 0x01;
    private ImagePreference preference; // hack
    private AlertDialog.Builder builder;
    private Button setButton;
    private Button selectButton;
    private Button deleteButton;
    private ImageView imageView;
    private Bitmap image;

    private void showErrorMessage(String title, String message) {
        builder.setTitle(title);
        builder.setMessage(message);
        builder.setPositiveButton(android.R.string.ok, null);
        builder.show();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_image);

        this.preference = ImagePreference.self; // hack, TODO: pass serialized bitmap in bundle
        this.image = this.preference.getImage();

        imageView = findViewById(R.id.selectedImage);

        builder = new AlertDialog.Builder(this);
        setButton = findViewById(R.id.SetButton);
        selectButton = findViewById(R.id.SelectButton);
        deleteButton = findViewById(R.id.DeleteButton);

        selectButton.setOnClickListener((View v) -> {
            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);
            intent.setType("image/*");
            startActivityForResult(intent, READ_IMAGE_REQUEST);
        });

        setButton.setOnClickListener((View v) -> {
            // persist your value here
            ImageActivity.this.preference.setImage(ImageActivity.this.image);
            ImageActivity.this.finish();
        });

        deleteButton.setOnClickListener((View v) -> {
            builder.setTitle("Confirm");
            builder.setMessage("Really remove image?");
            builder.setCancelable(false); // not necessary

            builder.setPositiveButton(R.string.yes, (DialogInterface dialog, int id) -> {
                ImageActivity.this.image = null;
                ImageActivity.this.updateImageView();
                dialog.cancel();
            });

            builder.setNegativeButton(R.string.no, (DialogInterface dialog, int id) -> {
                dialog.cancel();
            });

            // create dialog box
            AlertDialog alert = builder.create();
            alert.show();
        });

        updateImageView();
    }

    private void updateImageView() {
        imageView.setImageBitmap(image);
        //imageView.setImageDrawable(image);

        if (image == null) {
            deleteButton.setEnabled(false);
        } else {
            deleteButton.setEnabled(true);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == READ_IMAGE_REQUEST && resultCode == RESULT_OK) {
            if (data != null && data.getData() != null) {
                updateImage(data.getData());
            }
        }
    }

    void updateImage(Uri uri) {
        final int maxSize = 800;
        try {
            byte[] data = Utils.readFile(this, uri);
            Bitmap image = BitmapFactory.decodeByteArray(data, 0, data.length);

            if (image == null) {
                showErrorMessage("Error", "Not a supported image format: " + uri.getLastPathSegment());
                return;
            }

            final int inWidth = image.getWidth();
            final int inHeight = image.getHeight();

            int outWidth = 0;
            int outHeight = 0;

            if (inWidth > inHeight) {
                outWidth = maxSize;
                outHeight = (int) ((inHeight * maxSize) / (float) inWidth);
            } else {
                outHeight = maxSize;
                outWidth = (int) ((inWidth * maxSize) / (float) inHeight);
            }

            image = Bitmap.createScaledBitmap(image, outWidth, outHeight, false);
            ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
            boolean success = image.compress(Bitmap.CompressFormat.PNG, 0, byteStream);
            if (success) {
                //Log.d("ImageActivity", "image: " + inWidth + "/" + inHeight + ", compress.length: " + byteStream.toByteArray().length + ", base64: " + Base64.encodeToString(byteStream.toByteArray(), 0).length());
                this.image = image;
            } else {
                throw new Exception("Cannot compress image");
            }
            updateImageView();
        } catch (Exception e) {
            showErrorMessage("Error", e.toString());
        }
    }
}

File: app/src/main/kotlin/app/trigger/ImageActivity.kt
/*
* Copyright (C) 2025 The Trigger App Contributors
* SPDX-License-Identifier: GPL-3.0-or-later
*/

package app.trigger

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ImageView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import app.trigger.DoorStatus.StateCode
import java.io.ByteArrayOutputStream

class ImageActivity : AppCompatActivity() {
    private lateinit var door: Door
    private lateinit var stateCode: StateCode

    private lateinit var builder: AlertDialog.Builder
    private lateinit var setButton: Button
    private lateinit var selectButton: Button
    private lateinit var deleteButton: Button
    private lateinit var imageView: ImageView
    private var currentImage: Bitmap? = null

    private fun showErrorMessage(message: String) {
        builder.setTitle(R.string.error)
        builder.setMessage(message)
        builder.setPositiveButton(android.R.string.ok, null)
        builder.show()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_image)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        // currentDoor might still not be stored if it is a new one
        val currentDoor = SetupActivity.currentDoor
        val codeString = intent.getStringExtra("state_code") ?: ""
        if (currentDoor == null || codeString.isEmpty()) {
            // not expected to happen
            finish()
            return
        }

        door = currentDoor
        stateCode = StateCode.valueOf(codeString)
        currentImage = door.getStateImage(stateCode)

        imageView = findViewById(R.id.selectedImage)
        builder = AlertDialog.Builder(this)
        setButton = findViewById(R.id.SetButton)
        selectButton = findViewById(R.id.SelectButton)
        deleteButton = findViewById(R.id.DeleteButton)

        selectButton.setOnClickListener { _: View? ->
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "image/*"
            importImageLauncher.launch(intent)
        }

        setButton.setOnClickListener { _: View? ->
            // persist your value here
            door.setStateImage(stateCode, currentImage)
            finish()
        }

        deleteButton.setOnClickListener { _: View? ->
            builder.setTitle(R.string.confirm)
            builder.setMessage(R.string.dialog_really_remove_image)
            builder.setCancelable(false) // not necessary
            builder.setPositiveButton(R.string.yes) { dialog: DialogInterface, _: Int ->
                currentImage = null
                door.setStateImage(stateCode, null)
                updateImageView()
                dialog.cancel()
            }
            builder.setNegativeButton(R.string.no) { dialog: DialogInterface, _: Int -> dialog.cancel() }

            // create dialog box
            val alert = builder.create()
            alert.show()
        }

        updateImageView()
    }

    private fun updateImageView() {
        if (currentImage == null) {
            // show default image
            val defaultImageResource = when (stateCode) {
                StateCode.OPEN -> R.drawable.state_open
                StateCode.CLOSED -> R.drawable.state_closed
                StateCode.DISABLED -> R.drawable.state_disabled
                StateCode.UNKNOWN -> R.drawable.state_unknown
            }
            imageView.setImageBitmap(BitmapFactory.decodeResource(resources, defaultImageResource))
            deleteButton.isEnabled = false
        } else {
            imageView.setImageBitmap(currentImage)
            deleteButton.isEnabled = true
        }
    }

    private var importImageLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val intent = result.data ?: return@registerForActivityResult
            val uri = intent.data ?: return@registerForActivityResult
            updateImage(uri)
        }
    }

    private fun updateImage(uri: Uri) {
        val maxSize = 800
        try {
            val data = Utils.readFile(this, uri)
            var image = BitmapFactory.decodeByteArray(data, 0, data.size)
            if (image == null) {
                showErrorMessage("Not a supported image format: " + uri.lastPathSegment)
                return
            }
            val inWidth = image.width
            val inHeight = image.height
            var outWidth = 0
            var outHeight = 0
            if (inWidth > inHeight) {
                outWidth = maxSize
                outHeight = (inHeight * maxSize / inWidth.toFloat()).toInt()
            } else {
                outHeight = maxSize
                outWidth = (inWidth * maxSize / inHeight.toFloat()).toInt()
            }
            image = Bitmap.createScaledBitmap(image, outWidth, outHeight, false)
            val byteStream = ByteArrayOutputStream()
            val success = image.compress(Bitmap.CompressFormat.PNG, 0, byteStream)
            if (success) {
                //Log.d("ImageActivity", "image: " + inWidth + "/" + inHeight + ", compress.length: " + byteStream.toByteArray().length + ", base64: " + Base64.encodeToString(byteStream.toByteArray(), 0).length());
                currentImage = image
            } else {
                throw Exception("Cannot compress image")
            }
            updateImageView()
        } catch (e: Exception) {
            showErrorMessage(e.toString())
        }
    }

    companion object {
        private const val TAG = "ImageActivity"
    }
}
